forward/rewriting.lua

```diff
@@ -134,7 +135,64 @@ function Rewriting:process_packet(i, o)
    else
       l4_type = ip_header:next_header()
    end
-   if not (l4_type == L4_TCP or l4_type == L4_UDP) then
+   local ip_total_length = ip_header:total_length()
+   -- Check for IP fragmentation
+   local frag_off = ip_header:frag_off()
+   local mf = band(ip_header:flags(), IP_MF_FLAG) ~= 0
+   if frag_off ~= 0 or mf then
+      local src_port, dst_port = 0, 0
+      -- TODO: Redirect to spike backend pool.
+      P.free(p)
+      return
+   end
+   if l4_type == L4_GRE then
+      -- Start of IP fragment extraction and reassembly
+      -- Might want to extract this block into a separate function
+      local gre_class = ip_header:upper_layer()
+      local gre_header = datagram:parse_match(gre_class)
+      -- Checking the recursion control bit seems unnecessary here since
+      -- there is no other reason we would encounter a GRE header...?
```

Comment:
Can you expand on your reasoning here?  What prevents us from potentially forwarding packet fragments indefinitely?

Reply:
When a fragment arrives at the first spike, it should not have a GRE header. When it arrives at the second spike, it will have a GRE header. The existence of the GRE header would indicate to us that the fragment has already been forwarded once, causing the spike to process it in reassembly instead of forwarding it a second time.